### ISSUE TITLE
Python/C API abort when app build on 3.11 uses 3.10 environment

### ISSUE THREAD

**Issue by serban-seeq** [python/cpython#116139]
# Crash report

### What happened?

I have a C++ application using the Python/C API.

I recently updated it to use the 3.11 python.h and had to make the changes.

```c
   FN_Py_SetPythonHome(wszHome);
   FN_Py_SetProgramName(wszExe);
   FN_Py_Initialize();
   FN_PySys_SetArgvEx(1, argv, 0);
   FN_PyErr_Occurred();
```
changed to 
```c
   FN_PyConfig_InitPythonConfig(&config);
   FN_PyConfig_SetString(&config, &config.home, wszExe);
   FN_PyConfig_SetString(&config, &config.program_name, wszExe);
   FN_Py_InitializeFromConfig(&config);
   FN_PyConfig_SetArgv(&config, 1, argv);
   FN_PyErr_Occurred();
```

In the past I assumed there was some sort of backwards compatibility in place because I was able to run my code with an older python env in `wszHome` than the one I was building against.

I tested the new code various old versions of python. I was actually able to run python scripts inside my app on 3.8 and 3.9 but I got an unexpected `Access Violation Error` on 3.10

```
Unhandled exception at 0x00007FFD7A79D0DF (python310.dll) in omcore.exe: 0xC0000005: Access violation reading location 0x0000000000000058.
```

Is this even something that is expected to work and is the fact that 3.8 and 3.9 are working is just a happy accident?

[abort.zip](https://github.com/python/cpython/files/14452351/abort.zip)

### CPython versions tested on:

3.11

### Operating systems tested on:

Windows

**Comment by ronaldoussoren** [comment]
Building against Python 3.11 headers and running with Python 3.10 is not supported in general, the exception being the stable ABI (assuming you set ``Py_LIMITED_API`` to a value that includes support for 3.10.

In some cases building against newer headers happens to work, not always. In particular, the definition of C structs may change between versions of Python.

See also: https://docs.python.org/3/c-api/stable.html

**Comment by serban-seeq** [comment]
Reading though https://docs.python.org/3/c-api/stable.html I was hoping I would be able to keep using `Py_SetPythonHome`, set the `Py_LIMITED_API`  to something like `3.8` then compile against `3.11`, but I can't to that because `Py_SetPythonHome` (which is part of Limited API) has been deprecated in `3.11` and I have to use `PyConfig_SetString` which is not in Limited API.

So I guess if I want to keep the backwards AND forwards compatibility I can only compile up to version 3.10

Maybe im missing something...

**Comment by sunmy2019** [comment]
> So I guess if I want to keep the backwards AND forwards compatibility via `Py_LIMITED_API` I can only compile up to version 3.10

I guess so. Can you provide different compiled version to the user?

**Comment by serban-seeq** [comment]
Well the doc says about `Py_LIMITED_API` _The extension will work without recompilation with all Python 3 releases from the specified one onward_, but how does it achieve this if `Py_SetPythonHome` (which is part of Limited API) has been deprecated in 3.11 and I have to use PyConfig_SetString (which is not in Limited API) instead. So I take this as my application although hasn't changed in features, is no longer eligible for `Py_LIMITED_API` starting in 3.11 and onwards.

The original intent is we don't want to force the user to upgrade python when they upgrade our app. Its fine for now, I can build against 3.10 with `Py_LIMITED_API` but in the future in oct 2026 when 3.10 reached end of life, I wont want to build against 3.10 so I build against 3.11 and I have to drop `Py_LIMITED_API` so now I'm not guaranteed either forward or backward compatibility.

I guess this boils down to: a `Py_LIMITED_API` endpoint was deprecated and replaced with one not in `Py_LIMITED_API` and the documentation doesn't explicitly say what are the backwards/forwards compatibility implications of this.

**Comment by vstinner** [comment]
@serban-seeq: Please see my draft [PEP 741](https://peps.python.org/pep-0741/) "Python Configuration C API" which proposes adding a **stable ABI** to configure the Python initialization.

**Comment by pyscripter** [comment]
@vstinner A key problem is the stability of the PyConfig structure.  It has been changing in every version of python in some cases just by reordering the fields.   This is OK if you target a single version of python and you work in C/C++.   But if you want to be able to target multiple python versions with a single executable or you work in a different programming language (so you need to translate the headers) this is just a nightmare.  And with the removal of the old config API in 3.13, this is a problem that is impossible to avoid.

I am looking forward to PEP 741 being accepted and implemented to provide for a way out.

**Comment by vstinner** [comment]
> I just wonder why didn't you just add new fields at the end and insert dummy fields in the place of removed ones to keep some level of compatibility.

The problem is bigger than the PyConfig struct ABI.

And the reason is that the struct is excluded from the stable ABI, so there is no benefits to do what you say.

**Comment by picnixz** [comment]
Considering the OP's original issue has been (somewhat) resolved, I will close this one as `wont fix`. Note that a note on the limited API says (emphasis mine):

> Also note that the Limited API is not necessarily stable: compiling with Py_LIMITED_API with Python 3.8 means that the extension will run with Python 3.12, but it will not necessarily compile with Python 3.12. **In particular, parts of the Limited API may be deprecated and removed, provided that the Stable ABI stays stable.**

**Comment by vstinner** [comment]
> I am looking forward to [PEP 741](https://peps.python.org/pep-0741/) being accepted and implemented to provide for a way out.

PEP 741 was accepted and implemented in Python 3.14.
